streamer/proxy_node.py

```diff
@@ -0,0 +1,365 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A module that implements a simple proxy server for uploading packaged
+content to cloud storage providers (GCS and S3) and also any server that
+accepts PUT requests.
+"""
+
+import os
+import json
+import threading
+import urllib.parse
+from typing import Optional, Union, Dict
+from http.server import ThreadingHTTPServer, BaseHTTPRequestHandler
+from http.client import HTTPConnection, HTTPSConnection, CREATED, OK
+
+from streamer.node_base import ProcessStatus, ThreadedNodeBase
+from streamer.util import RequestBodyAsFileIO
+
+
+# Protocols we have classes for to handle them.
+SUPPORTED_PROTOCOLS = ['http', 'https', 'gs', 's3']
+
+
+class RequestHandler(BaseHTTPRequestHandler):
+  """A request handler that processes the PUT requests coming from
+  shaka packager and pushes them to the destination.
+  """
+
+  def __init__(self, conn: Union[HTTPConnection, HTTPSConnection],
+               extra_headers: Dict[str, str], base_path: str, param_query: str,
+               temp_dir: Optional[str], *args, **kwargs):
+
+    self._conn = conn
+    # Extra headers to add when sending the request to the host
+    # using `self._conn`.
+    self._extra_headers = extra_headers
+    # The base path that will be prepended to the path of the handled request
+    # before forwarding the request to the host using `self._conn`.
+    self._base_path = base_path
+    # Parameters and query string to send in the url with each forwarded request.
+    self._params_and_querystring = param_query
+    self._temp_dir = temp_dir
+    # Call `super().__init__()` last because this call is what handles the
+    # actual request and we need the variables defined above to handle
+    # this request.
+    super().__init__(*args, **kwargs)
+
+  def do_PUT(self):
+    """do_PUT will handle the PUT requests coming from shaka packager."""
+
+    headers = {}
+    # Use the same headers for requesting.
+    for k, v in self.headers.items():
+      if k.lower() != 'host':
+        headers[k] = v
+    # Add the extra headers, this might contain an access token for instance.
+    for k, v in self._extra_headers.items():
+      headers[k] = v
+
+    # Don't chunk by default, as the request body is already chunked
+    # or we have a content-length header which means we are not using
+    # chunked transfer-encoding.
+    encode_chunked = False
+    content_length = self.headers['Content-Length']
+    # Store the manifest files locally in `self._temp_dir`.
+    if self._temp_dir is not None and self.path.endswith(('.mpd', '.m3u8')):
+      body = self._write_body_and_get_file_io()
+      if content_length == None:
+        # We need to re-chunk it again as we wrote it as a whole
+        # to the filesystem.
+        encode_chunked = True
+    else:
+      content_length = content_length and int(content_length)
+      body = RequestBodyAsFileIO(self.rfile, content_length)
```

Comment:
I'm surprised that any of this is needed: either the file IO wrapper or the temporary writing of chunks to the FS.  Also, writing chunks to the FS means the output is no longer being streamed to chunk-capable clouds like Amazon's.  Ideally, we would not buffer a whole segment before transmitting it.

I'm going to experiment with this a bit locally and see what I can do.

Reply:
In here we will write only the manifests to the FS so we can perform period concatenation on them (note: the period concatenation is carried out by the streamer and not packager, if the manifests are pushed to the remote url straight away, we won't be able to perform period concatenation). We won't write media segments to the FS though.

But, re-looking at this, I think I over complicated it (to avoid introducing new dependencies, I believe). I will look into a package that supports streaming to carry this out (probably [`requests`](https://pypi.org/project/requests/)).